File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuListUtils.scala
/*
 * Copyright (c) 2021, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids

import java.util.Optional

import ai.rapids.cudf.{ColumnView, DType}

/**
 * Provide a set of APIs to manipulate array/list columns in common ways.
 */
object GpuListUtils extends Arm {
  /**
   * Replace the data column in a LIST column. This will keep the same offsets and validity
   * of the listColumn.  This returns a view so it is the responsibility of the caller to keep
   * both listCol and newDataCol alive longer than the returned ColumnView.
   * @param listCol the list column to use as a template
   * @param newDataCol the new data column.
   * @return a new ColumnView.
   * @throws IllegalArgumentException if data column does not match the original data column in
   *                                  size.
   */
  def replaceListDataColumn(
      listCol: ColumnView,
      newDataCol: ColumnView): ColumnView = {
    assert(DType.LIST.equals(listCol.getType))
    withResource(listCol.getChildColumnView(0)) { dataCol =>
      if (dataCol.getRowCount != newDataCol.getRowCount) {
        throw new IllegalArgumentException("Mismatch in the number of rows in the data columns")
      }
    }
    withResource(listCol.getOffsets) { offsets =>
      withResource(listCol.getValid) { validity =>
        new ColumnView(DType.LIST, listCol.getRowCount,
          Optional.of[java.lang.Long](listCol.getNullCount), validity, offsets,
          Array[ColumnView](newDataCol))
      }
    }
  }
}

File: sql-plugin-columnar/src/main/java/com/nvidia/spark/rapids/GpuListUtils.java
/*
 * Copyright (c) 2021-2026, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids;

import java.util.Optional;

import ai.rapids.cudf.ColumnView;
import ai.rapids.cudf.DType;
import ai.rapids.cudf.BaseDeviceMemoryBuffer;

/** Provides APIs to manipulate array/list columns in common ways. */
public final class GpuListUtils {
  private GpuListUtils() {}

  /**
   * Replace the data column in a LIST column. This keeps the same offsets and validity
   * of the list column. This returns a view, so the caller is responsible for keeping
   * both {@code listCol} and {@code newDataCol} alive longer than the returned view.
   *
   * @param listCol the list column to use as a template
   * @param newDataCol the new data column
   * @return a new ColumnView
   * @throws IllegalArgumentException if the data column does not match the original data column
   *     in size
   */
  public static ColumnView replaceListDataColumnAsView(
      ColumnView listCol, ColumnView newDataCol) {
    assert DType.LIST.equals(listCol.getType());
    try (ColumnView dataCol = listCol.getChildColumnView(0)) {
      if (dataCol.getRowCount() != newDataCol.getRowCount()) {
        throw new IllegalArgumentException("Mismatch in the number of rows in the data columns");
      }
    }
    try (BaseDeviceMemoryBuffer offsets = listCol.getOffsets();
         BaseDeviceMemoryBuffer validity = listCol.getValid()) {
      return new ColumnView(
          DType.LIST,
          listCol.getRowCount(),
          Optional.of(listCol.getNullCount()),
          validity,
          offsets,
          new ColumnView[] { newDataCol });
    }
  }
}
